ConnectionManager not creating new connections when needed [JIRA: CLIENTS-586]

Hi,

The ConnectionManager is not creating new connections when the existing ones are supposed to be busy. I haven't been able to look into this deeper but it can be reproduced this way:

- Cluster using one node, MinConnections: 10 MaxConnections: 200
- Process launched, I can see 10 connections using netstat that are killed after a few seconds (#23)
- After that I run some benchmarks that trigger a lot of concurrent calls to Riak, but checking netstat no more than one connection is ever created, affecting the application performance considerably

I was expecting it to create connections up to MaxConnections as needed. I'll try to run some more tests next week to give you some more information.

Thanks.-

Hi Luke,

I've doing some more tests and everything seems to be working fine. The errors I mentioned to you I was getting only happen when talking to Riak through Haproxy, and the problem could be anything related to my local configuration or environment (I don't remember experimenting similar issues when using the unofficial goriakpbc client a few weeks ago in the same environment, though). When talking directly to Riak all works perfectly fine.

When talking to Riak through Haproxy I start getting errors like the ones below until all connections are killed (the 10 set as minimum initially). Interestingly, after a few seconds new connections are created again and everything seems to start working fine (after the initial storm of errors). I mention this just as a curiosity.

2015/09/14 10:44:38 [DEBUG] [Node] (172.17.8.101:8087|10|9) - executing command 'UpdateMap'
2015/09/14 10:44:38 [DEBUG] [Connection] execute command: UpdateMap
2015/09/14 10:44:38 [DEBUG] [DefaultNodeManager] executed 'UpdateMap' on node '172.17.8.101:8087|9|9', err 'EOF'
2015/09/14 10:44:38 [DEBUG] [Cluster] executed command 'UpdateMap': re-try due to error 'EOF'
2015/09/14 10:44:38 [DEBUG] [commandImpl] remainingTries: 0
2015/09/14 10:44:38 [ERROR] [Async] ClientError|[Cluster] all retries exhausted and/or no nodes available to execute command|InnerError|EOF
2015/09/14 10:44:38 [DEBUG] [Node] (172.17.8.101:8087|9|9) starting health check routine

Thanks again.-

--- a/cluster.go
+++ b/cluster.go
@@ -76,7 +76,6 @@ func NewCluster(options *ClusterOptions) (c *Cluster, err error) {
 		c = nil
 		return
 	}
-	c.nodeManager.Init(c.nodes)
 
 	if options.QueueMaxDepth > 0 {
 		if options.QueueExecutionInterval == 0 {
@@ -141,7 +140,7 @@ func (c *Cluster) Stop() (err error) {
 		c.commandQueueTicker.Stop()
 		qc := c.cq.count()
 		if qc > 0 {
-			logWarn("[Cluster]", "commands in queue during shutdown: ", qc)
+			logWarn("[Cluster]", "commands in queue during shutdown: %d", qc)
 			var f = func(v interface{}) (bool, bool) {
 				if v == nil {
 					return true, false
@@ -176,7 +175,6 @@ func (c *Cluster) Stop() (err error) {
 	}
 
 	if allStopped {
-		c.nodeManager.Stop()
 		c.setState(clusterShutdown)
 		logDebug("[Cluster]", "cluster shut down")
 	} else {
@@ -277,7 +275,7 @@ func (c *Cluster) execute(async *Async) {
 		if err = c.stateCheck(clusterRunning); err != nil {
 			break
 		}
-		executed, err = c.nodeManager.ExecuteOnNode(command, command.getLastNode())
+		executed, err = c.nodeManager.ExecuteOnNode(c.nodes, command, command.getLastNode())
 		// NB: do *not* call command.onError here as it will have been called in connection
 		if executed {
 			// NB: "executed" means that a node sent the data to Riak and received a response

--- a/cluster_i_test.go
+++ b/cluster_i_test.go
@@ -70,10 +70,10 @@ func TestExecuteCommandOnCluster(t *testing.T) {
 }
 
 func TestExecuteConcurrentCommandsOnCluster(t *testing.T) {
-	count := uint16(8)
+	maxCount := uint16(8)
 	nodeOpts := &NodeOptions{
-		MinConnections: count,
-		MaxConnections: count,
+		MinConnections: 1,
+		MaxConnections: 4,
 		RemoteAddress:  getRiakAddress(),
 	}
 
@@ -106,9 +106,8 @@ func TestExecuteConcurrentCommandsOnCluster(t *testing.T) {
 		t.Error(err.Error())
 	}
 
-	pingChan := make(chan *PingCommand, count)
-	for i := uint16(0); i < count; i++ {
-		logDebug("[TestExecuteConcurrentCommandsOnCluster]", "i: %d", i)
+	pingChan := make(chan *PingCommand)
+	for i := uint16(0); i < maxCount; i++ {
 		go func() {
 			command := &PingCommand{}
 			if err := cluster.Execute(command); err != nil {
@@ -119,16 +118,14 @@ func TestExecuteConcurrentCommandsOnCluster(t *testing.T) {
 	}
 
 	j := uint16(0)
-	for i := uint16(0); i < count; i++ {
-		select {
-		case pingCommand := <-pingChan:
-			if expected, actual := true, pingCommand.Success(); expected != actual {
-				t.Errorf("expected %v, got %v", expected, actual)
-			}
-			j++
+	for i := uint16(0); i < maxCount; i++ {
+		pingCommand := <-pingChan
+		if expected, actual := true, pingCommand.Success(); expected != actual {
+			t.Errorf("expected %v, got %v", expected, actual)
 		}
+		j++
 	}
-	if expected, actual := count, j; expected != actual {
+	if expected, actual := maxCount, j; expected != actual {
 		t.Errorf("expected %v, got %v", expected, actual)
 	}
 }

--- a/command_i_test.go
+++ b/command_i_test.go
@@ -22,7 +22,6 @@ func TestPing(t *testing.T) {
 		remoteAddress:  addr,
 		connectTimeout: time.Second * 5,
 		requestTimeout: time.Millisecond * 500,
-		healthCheck:    &PingCommand{},
 	}
 	if conn, err = newConnection(opts); err == nil {
 		if err = conn.connect(); err == nil {

--- a/connection.go
+++ b/connection.go
@@ -29,7 +29,6 @@ type connectionOptions struct {
 	remoteAddress  *net.TCPAddr
 	connectTimeout time.Duration
 	requestTimeout time.Duration
-	healthCheck    Command
 	authOptions    *AuthOptions
 }
 
@@ -45,7 +44,6 @@ type connection struct {
 	conn           net.Conn
 	connectTimeout time.Duration
 	requestTimeout time.Duration
-	healthCheck    Command
 	authOptions    *AuthOptions
 	sizeBuf        []byte
 	dataBuf        []byte
@@ -72,7 +70,6 @@ func newConnection(options *connectionOptions) (*connection, error) {
 		addr:           options.remoteAddress,
 		connectTimeout: options.connectTimeout,
 		requestTimeout: options.requestTimeout,
-		healthCheck:    options.healthCheck,
 		authOptions:    options.authOptions,
 		sizeBuf:        make([]byte, 4),
 		dataBuf:        make([]byte, defaultInitBuffer),
@@ -100,13 +97,6 @@ func (c *connection) connect() (err error) {
 			return
 		}
 		c.setState(connActive)
-		if c.healthCheck != nil {
-			if err = c.execute(c.healthCheck); err != nil || !c.healthCheck.Success() {
-				c.setState(connInactive)
-				logError("[Connection]", "initial health check error: '%s'", err.Error())
-				c.close()
-			}
-		}
 	}
 	return
 }

--- a/connection_i_test.go
+++ b/connection_i_test.go
@@ -134,50 +134,7 @@ func TestConnectionTimeout(t *testing.T) {
 	}
 }
 
-func TestHealthCheckFail(t *testing.T) {
-	var onConn = func(c net.Conn) bool {
-		handleClientMessageWithRiakError(t, c, 1, nil)
-		return true
-	}
-	o := &testListenerOpts{
-		test:   t,
-		host:   "127.0.0.1",
-		port:   1339,
-		onConn: onConn,
-	}
-	tl := newTestListener(o)
-	defer tl.stop()
-	tl.start()
-
-	addr, err := net.ResolveTCPAddr("tcp4", tl.addr)
-	if err != nil {
-		t.Error(err.Error())
-	}
-
-	opts := &connectionOptions{
-		remoteAddress:  addr,
-		connectTimeout: thirtySeconds,
-		healthCheck:    &PingCommand{},
-	}
-
-	if conn, err := newConnection(opts); err == nil {
-		if err := conn.connect(); err == nil {
-			t.Error("expected to see error")
-		} else {
-			if riakError, ok := err.(RiakError); ok == true {
-				if expected, actual := "RiakError|1|this is an error", riakError.Error(); expected != actual {
-					t.Errorf("expected %v, got %v", expected, actual)
-				}
-			} else {
-				t.Error("expected to see Riak error, got:", err)
-			}
-		}
-	} else {
-		t.Error(err)
-	}
-}
-
-func TestHealthCheckSuccess(t *testing.T) {
+func TestConnectionSuccess(t *testing.T) {
 	o := &testListenerOpts{
 		test: t,
 		host: "127.0.0.1",
@@ -195,7 +152,6 @@ func TestHealthCheckSuccess(t *testing.T) {
 	opts := &connectionOptions{
 		remoteAddress:  addr,
 		connectTimeout: thirtySeconds,
-		healthCheck:    &PingCommand{},
 	}
 
 	if conn, err := newConnection(opts); err == nil {

--- a/connection_manager.go
+++ b/connection_manager.go
@@ -17,37 +17,38 @@ const (
 )
 
 type connectionManagerOptions struct {
-	addr           *net.TCPAddr
-	stopChan       chan bool
-	minConnections uint16
-	maxConnections uint16
-	idleTimeout    time.Duration
-	connectTimeout time.Duration
-	requestTimeout time.Duration
-	authOptions    *AuthOptions
+	addr                   *net.TCPAddr
+	minConnections         uint16
+	maxConnections         uint16
+	idleExpirationInterval time.Duration
+	idleTimeout            time.Duration
+	connectTimeout         time.Duration
+	requestTimeout         time.Duration
+	authOptions            *AuthOptions
 }
 
 type connectionManager struct {
-	addr            *net.TCPAddr
-	minConnections  uint16
-	maxConnections  uint16
-	idleTimeout     time.Duration
-	connectTimeout  time.Duration
-	requestTimeout  time.Duration
-	authOptions     *AuthOptions
-	stopChan        chan bool
-	q               *queue
-	expireTicker    *time.Ticker
-	connectionCount uint16
+	addr                   *net.TCPAddr
+	minConnections         uint16
+	maxConnections         uint16
+	idleExpirationInterval time.Duration
+	idleTimeout            time.Duration
+	connectTimeout         time.Duration
+	requestTimeout         time.Duration
+	authOptions            *AuthOptions
+	stopChan               chan struct{}
+	q                      *queue
+	expireTicker           *time.Ticker
+	connectionCount        uint16
 	sync.RWMutex
 	stateData
 }
 
 var (
-	ErrConnectionManagerRequiresOptions     = newClientError("[connectionManager] new manager requires options", nil)
-	ErrConnectionManagerRequiresAddress     = newClientError("[connectionManager] new manager requires non-nil address", nil)
-	ErrConnectionManagerRequiresStopChannel = newClientError("[connectionManager] new manager requires non-nil stop channel", nil)
-	ErrConnMgrAllConnectionsInUse           = newClientError("[connectionManager] all connections in use at max connections reached", nil)
+	ErrConnectionManagerRequiresOptions         = newClientError("[connectionManager] new manager requires options", nil)
+	ErrConnectionManagerRequiresAddress         = newClientError("[connectionManager] new manager requires non-nil address", nil)
+	ErrConnectionManagerMaxMustBeGreaterThanMin = newClientError("[connectionManager] new connection manager maxConnections must be greater than minConnections", nil)
+	ErrConnMgrAllConnectionsInUse               = newClientError("[connectionManager] all connections in use / max connections reached", nil)
 )
 
 func newConnectionManager(options *connectionManagerOptions) (*connectionManager, error) {
@@ -57,14 +58,17 @@ func newConnectionManager(options *connectionManagerOptions) (*connectionManager
 	if options.addr == nil {
 		return nil, ErrConnectionManagerRequiresAddress
 	}
-	if options.stopChan == nil {
-		return nil, ErrConnectionManagerRequiresStopChannel
-	}
 	if options.minConnections == 0 {
 		options.minConnections = defaultMinConnections
 	}
 	if options.maxConnections == 0 {
 		options.maxConnections = defaultMaxConnections
+	}
+	if options.minConnections > options.maxConnections {
+		return nil, ErrConnectionManagerMaxMustBeGreaterThanMin
+	}
+	if options.idleExpirationInterval == 0 {
+		options.idleExpirationInterval = defaultIdleExpirationInterval
 	}
 	if options.idleTimeout == 0 {
 		options.idleTimeout = defaultIdleTimeout
@@ -76,15 +80,16 @@ func newConnectionManager(options *connectionManagerOptions) (*connectionManager
 		options.requestTimeout = defaultRequestTimeout
 	}
 	cm := &connectionManager{
-		addr:           options.addr,
-		minConnections: options.minConnections,
-		maxConnections: options.maxConnections,
-		idleTimeout:    options.idleTimeout,
-		connectTimeout: options.connectTimeout,
-		requestTimeout: options.requestTimeout,
-		authOptions:    options.authOptions,
-		stopChan:       options.stopChan,
-		q:              newQueue(options.maxConnections),
+		addr:                   options.addr,
+		minConnections:         options.minConnections,
+		maxConnections:         options.maxConnections,
+		idleExpirationInterval: options.idleExpirationInterval,
+		idleTimeout:            options.idleTimeout,
+		connectTimeout:         options.connectTimeout,
+		requestTimeout:         options.requestTimeout,
+		authOptions:            options.authOptions,
+		stopChan:               make(chan struct{}),
+		q:                      newQueue(options.maxConnections),
 	}
 	cm.initStateData("connMgrError", "connMgrCreated", "connMgrRunning", "connMgrShuttingDown", "connMgrShutdown")
 	cm.setState(cmCreated)
@@ -92,7 +97,7 @@ func newConnectionManager(options *connectionManagerOptions) (*connectionManager
 }
 
 func (cm *connectionManager) String() string {
-	return fmt.Sprintf("%v|%d", cm.addr, cm.count())
+	return fmt.Sprintf("%v", cm.addr)
 }
 
 func (cm *connectionManager) start() error {
@@ -100,14 +105,14 @@ func (cm *connectionManager) start() error {
 		return err
 	}
 	for i := uint16(0); i < cm.minConnections; i++ {
-		conn, err := cm.create(nil)
+		conn, err := cm.create()
 		if err == nil {
 			cm.put(conn)
 		} else {
 			logErr("[connectionManager]", err)
 		}
 	}
-	cm.expireTicker = time.NewTicker(fiveSeconds)
+	cm.expireTicker = time.NewTicker(cm.idleExpirationInterval)
 	go cm.expireConnections()
 	cm.setState(cmRunning)
 	return nil
@@ -121,6 +126,7 @@ func (cm *connectionManager) stop() error {
 	logDebug("[connectionManager]", "shutting down")
 
 	cm.setState(cmShuttingDown)
+	close(cm.stopChan)
 	cm.expireTicker.Stop()
 
 	if cm.count() != cm.q.count() {
@@ -163,33 +169,40 @@ func (cm *connectionManager) count() uint16 {
 	return cm.connectionCount
 }
 
-func (cm *connectionManager) create(hc Command) (*connection, error) {
+func (cm *connectionManager) create() (*connection, error) {
 	if !cm.isStateLessThan(cmShuttingDown) {
 		return nil, nil
 	}
+
 	cm.Lock()
 	defer cm.Unlock()
-	if cm.connectionCount < cm.maxConnections {
-		opts := &connectionOptions{
-			remoteAddress:  cm.addr,
-			connectTimeout: cm.connectTimeout,
-			requestTimeout: cm.requestTimeout,
-			authOptions:    cm.authOptions,
-			healthCheck:    hc,
-		}
-		conn, err := newConnection(opts)
-		if err != nil {
-			return nil, err
-		}
-		err = conn.connect()
-		if err != nil {
-			return nil, err
-		}
-		cm.connectionCount++
-		return conn, nil
-	} else {
+
+	if cm.connectionCount >= cm.maxConnections {
 		return nil, ErrConnMgrAllConnectionsInUse
 	}
+
+	conn, err := cm.createConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	cm.connectionCount++
+	return conn, nil
+}
+
+func (cm *connectionManager) createConnection() (*connection, error) {
+	opts := &connectionOptions{
+		remoteAddress:  cm.addr,
+		connectTimeout: cm.connectTimeout,
+		requestTimeout: cm.requestTimeout,
+		authOptions:    cm.authOptions,
+	}
+	conn, err := newConnection(opts)
+	if err != nil {
+		return nil, err
+	}
+	err = conn.connect()
+	return conn, err
 }
 
 func (cm *connectionManager) get() (*connection, error) {
@@ -218,7 +231,7 @@ func (cm *connectionManager) get() (*connection, error) {
 	}
 
 	// NB: if we get here, there were no available connections
-	return cm.create(nil)
+	return cm.create()
 }
 
 func (cm *connectionManager) put(conn *connection) error {
@@ -261,6 +274,7 @@ func (cm *connectionManager) expireConnections() {
 
 			expiredCount := uint16(0)
 			now := time.Now()
+
 			var f = func(v interface{}) (bool, bool) {
 				if v == nil {
 					// connection pool is empty
@@ -270,16 +284,18 @@ func (cm *connectionManager) expireConnections() {
 					return true, true
 				}
 				conn := v.(*connection)
-				// expire connection if not available or if it has passed idle timeout
-				if !conn.available() || (now.Sub(conn.lastUsed) >= cm.idleTimeout) {
-					cm.Lock()
-					cm.connectionCount--
-					cm.Unlock()
-					if err := conn.close(); err != nil {
-						logErr("[connectionManager]", err)
+				cm.Lock()
+				defer cm.Unlock()
+				if cm.connectionCount > cm.minConnections {
+					// expire connection if not available or if it has passed idle timeout
+					if !conn.available() || (now.Sub(conn.lastUsed) >= cm.idleTimeout) {
+						cm.connectionCount--
+						if err := conn.close(); err != nil {
+							logErr("[connectionManager]", err)
+						}
+						expiredCount++
+						return false, false
 					}
-					expiredCount++
-					return false, false
 				}
 				return false, true
 			}

--- a/connection_manager_i_test.go
+++ b/connection_manager_i_test.go
@@ -1,0 +1,53 @@
+// +build integration
+
+package riak
+
+import (
+	"net"
+	"testing"
+	"time"
+)
+
+func TestConnectionManagerDoesNotExpirePastMinConnections(t *testing.T) {
+	minConnections := uint16(10)
+
+	o := &testListenerOpts{
+		test: t,
+		host: "127.0.0.1",
+		port: 13340,
+	}
+	tl := newTestListener(o)
+	tl.start()
+	defer tl.stop()
+
+	addr, err := net.ResolveTCPAddr("tcp", "127.0.0.1:13340")
+
+	cmopts := &connectionManagerOptions{
+		addr:                   addr,
+		minConnections:         minConnections,
+		maxConnections:         20,
+		idleExpirationInterval: time.Millisecond * 500,
+		idleTimeout:            time.Millisecond * 10,
+	}
+
+	cm, err := newConnectionManager(cmopts)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = cm.start()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < 10; i++ {
+		time.Sleep(time.Millisecond * 250)
+		if actual, expected := cm.connectionCount, minConnections; actual != expected {
+			t.Errorf("got: %v, expected: %v", actual, expected)
+		}
+	}
+
+	err = cm.stop()
+	if err != nil {
+		t.Error(err)
+	}
+}

--- a/consts.go
+++ b/consts.go
@@ -13,6 +13,7 @@ const (
 	defaultRemotePort             = uint16(8087)
 	defaultMinConnections         = uint16(1)
 	defaultMaxConnections         = uint16(8096)
+	defaultIdleExpirationInterval = fiveSeconds
 	defaultIdleTimeout            = threeSeconds
 	defaultConnectTimeout         = thirtySeconds
 	defaultRequestTimeout         = fiveSeconds

--- a/globals_i_test.go
+++ b/globals_i_test.go
@@ -184,7 +184,9 @@ func readClientMessage(c net.Conn) (err error) {
 		}
 		// logDebug("[readClientMessage]", "read message '%v', count '%d'", data, count)
 	} else {
-		err = errors.New(fmt.Sprintf("[readClientMessage] error reading command size into sizeBuf: count %d, err %s", count, err))
+		if err != io.EOF {
+			err = errors.New(fmt.Sprintf("[readClientMessage] error reading command size into sizeBuf: count %d, err %s", count, err))
+		}
 	}
 	return
 }

--- a/node.go
+++ b/node.go
@@ -37,7 +37,7 @@ type Node struct {
 	healthCheckInterval time.Duration
 	healthCheckBuilder  CommandBuilder
 	authOptions         *AuthOptions
-	stopChan            chan bool
+	stopChan            chan struct{}
 	// Health Check
 	expireTicker *time.Ticker
 	// Connections
@@ -71,9 +71,8 @@ func NewNode(options *NodeOptions) (*Node, error) {
 	var resolvedAddress *net.TCPAddr
 	resolvedAddress, err = net.ResolveTCPAddr("tcp", options.RemoteAddress)
 	if err == nil {
-		stopChan := make(chan bool)
 		n := &Node{
-			stopChan:            stopChan,
+			stopChan:            make(chan struct{}),
 			addr:                resolvedAddress,
 			healthCheckInterval: options.HealthCheckInterval,
 			healthCheckBuilder:  options.HealthCheckBuilder,
@@ -82,7 +81,6 @@ func NewNode(options *NodeOptions) (*Node, error) {
 
 		connMgrOpts := &connectionManagerOptions{
 			addr:           resolvedAddress,
-			stopChan:       stopChan,
 			minConnections: options.MinConnections,
 			maxConnections: options.MaxConnections,
 			idleTimeout:    options.IdleTimeout,
@@ -105,7 +103,7 @@ func NewNode(options *NodeOptions) (*Node, error) {
 // String returns a formatted string including the remoteAddress for the Node and its current
 // connection count
 func (n *Node) String() string {
-	return fmt.Sprintf("%v|%d", n.addr, n.cm.count())
+	return fmt.Sprintf("%v|%d|%d", n.addr, n.cm.count(), n.cm.q.count())
 }
 
 // Start opens a connection with Riak at the configured remoteAddress and adds the connections to the
@@ -133,7 +131,6 @@ func (n *Node) stop() error {
 	logDebug("[Node]", "(%v) shutting down.", n)
 
 	n.setState(nodeShuttingDown)
-	n.stopChan <- true
 	close(n.stopChan)
 
 	err := n.cm.stop()
@@ -143,6 +140,7 @@ func (n *Node) stop() error {
 		logDebug("[Node]", "(%v) shut down.", n)
 	} else {
 		n.setState(nodeError)
+		logErr("[Node]", err)
 	}
 
 	return err
@@ -250,7 +248,7 @@ func (n *Node) ensureHealthCheckCanContinue() bool {
 
 func (n *Node) healthCheck() {
 
-	logDebug("[Node]", "(%v) running health check", n)
+	logDebug("[Node]", "(%v) starting health check routine", n)
 
 	healthCheckTicker := time.NewTicker(n.healthCheckInterval)
 	defer healthCheckTicker.Stop()
@@ -265,12 +263,20 @@ func (n *Node) healthCheck() {
 			case t := <-healthCheckTicker.C:
 				if n.ensureHealthCheckCanContinue() {
 					logDebug("[Node]", "(%v) running health check at %v", n, t)
-					if conn, err := n.cm.create(healthCheckCommand); err != nil {
-						logDebug("[Node]", "(%v) failed healthcheck, err: %v", n, err)
+					conn, cerr := n.cm.createConnection()
+					if cerr != nil {
+						conn.close()
+						logDebug("[Node]", "(%v) failed healthcheck in create(false), err: %v", n, cerr)
 					} else {
-						n.cm.put(conn)
-						n.setState(nodeRunning)
-						logDebug("[Node]", "(%v) healthcheck success", n)
+						defer conn.close()
+						if n.ensureHealthCheckCanContinue() {
+							if hcerr := conn.execute(healthCheckCommand); hcerr != nil || !healthCheckCommand.Success() {
+								logDebug("[Node]", "(%v) failed healthcheck in execute, err: %v", n, hcerr)
+							} else {
+								n.setState(nodeRunning)
+								logDebug("[Node]", "(%v) healthcheck success", n)
+							}
+						}
 						return
 					}
 				}

--- a/node_i_test.go
+++ b/node_i_test.go
@@ -66,9 +66,6 @@ func TestCreateNodeWithOptionsAndStart(t *testing.T) {
 		if conn.addr.Zone != "" {
 			t.Errorf("expected empty zone, got: %s", string(conn.addr.Zone))
 		}
-		if conn.healthCheck != nil {
-			t.Error("expected nil conn.healthCheck")
-		}
 		if expected, actual := conn.connectTimeout, opts.ConnectTimeout; expected != actual {
 			t.Errorf("expected %v, got: %v", expected, actual)
 		}

--- a/node_manager.go
+++ b/node_manager.go
@@ -1,89 +1,66 @@
 package riak
 
 import (
-	"time"
-
-	"github.com/basho/backoff"
+	"sync"
 )
 
 // NodeManager enforces the structure needed to if going to implement your own NodeManager
 type NodeManager interface {
-	Init(nodes []*Node) error
-	ExecuteOnNode(command Command, previousNode *Node) (bool, error)
-	Stop()
+	ExecuteOnNode(nodes []*Node, command Command, previousNode *Node) (bool, error)
 }
 
 var ErrDefaultNodeManagerRequiresNode = newClientError("Must pass at least one node to default node manager", nil)
 
 type defaultNodeManager struct {
-	qsz uint16
-	q   *queue
-	b   *backoff.Backoff
-}
-
-func (nm *defaultNodeManager) Init(nodes []*Node) error {
-	if nodes == nil {
-		panic("[defaultNodeManager] nil nodes argument")
-	}
-	if len(nodes) == 0 || nodes[0] == nil {
-		return ErrDefaultNodeManagerRequiresNode
-	}
-	nm.qsz = uint16(len(nodes))
-	nm.q = newQueue(nm.qsz)
-	nm.b = &backoff.Backoff{
-		Min:    10 * time.Millisecond,
-		Max:    100 * time.Millisecond,
-		Jitter: true,
-	}
-	for _, n := range nodes {
-		nm.q.enqueue(n)
-	}
-	return nil
+	nodeIndex int
+	sync.RWMutex
 }
 
 // ExecuteOnNode selects a Node from the pool and executes the provided Command on that Node. The
 // defaultNodeManager uses a simple round robin approach to distributing load
-func (nm *defaultNodeManager) ExecuteOnNode(command Command, previous *Node) (bool, error) {
-	var err error
-	var executed bool = false
-
-	i := uint16(0)
-	var f = func(v interface{}) (bool, bool) {
-		if v == nil {
-			// pool is empty now, re-try
-			d := nm.b.Duration()
-			logDebug("[DefaultNodeManager]", "pool is empty, sleeping for '%v'", d)
-			time.Sleep(d)
-			return false, false
-		}
-		i++
-		node := v.(*Node)
-
-		// don't try the same node twice in a row if we have multiple nodes
-		if nm.qsz > 1 && previous != nil && previous == node {
-			return false, true
-		}
-
-		if executed, err = node.execute(command); executed == true {
-			logDebug("[DefaultNodeManager]", "executed '%s' on node '%s', err '%v'", command.Name(), node, err)
-			return true, true
-		}
-
-		if i == nm.qsz {
-			logDebug("[DefaultNodeManager]", "tried all nodes to execute '%s'", command.Name())
-			return true, true
-		}
-
-		return false, true
+func (nm *defaultNodeManager) ExecuteOnNode(nodes []*Node, command Command, previous *Node) (bool, error) {
+	if nodes == nil {
+		panic("[defaultNodeManager] nil nodes argument")
+	}
+	if len(nodes) == 0 || nodes[0] == nil {
+		return false, ErrDefaultNodeManagerRequiresNode
 	}
 
-	if ierr := nm.q.iterate(f); ierr != nil {
-		logErr("[DefaultNodeManager] iteration error", ierr)
+	var err error
+	executed := false
+
+	nm.RLock()
+	startingIndex := nm.nodeIndex
+	nm.RUnlock()
+
+	for {
+		nm.Lock()
+		if nm.nodeIndex >= len(nodes) {
+			nm.nodeIndex = 0
+		}
+		node := nodes[nm.nodeIndex]
+		nm.nodeIndex++
+		nm.Unlock()
+
+		// don't try the same node twice in a row if we have multiple nodes
+		if len(nodes) > 1 && previous != nil && previous == node {
+			continue
+		}
+
+		executed, err = node.execute(command)
+		if executed == true {
+			logDebug("[DefaultNodeManager]", "executed '%s' on node '%s', err '%v'", command.Name(), node, err)
+			break
+		}
+
+		nm.RLock()
+		if startingIndex == nm.nodeIndex {
+			nm.RUnlock()
+			// logDebug("[DefaultNodeManager]", "startingIndex %d nm.nodeIndex %d", startingIndex, nm.nodeIndex)
+			break
+		}
+		nm.RUnlock()
 	}
 
 	return executed, err
-}
-
-func (nm *defaultNodeManager) Stop() {
-	nm.q.destroy()
 }

--- a/node_test.go
+++ b/node_test.go
@@ -21,7 +21,7 @@ func TestCreateNodeWithOptions(t *testing.T) {
 	if err != nil {
 		t.Error(err.Error())
 	}
-	if expected, actual := "8.8.8.8:1234|0", node.String(); expected != actual {
+	if expected, actual := "8.8.8.8:1234|0|0", node.String(); expected != actual {
 		t.Errorf("expected %v, got %v", expected, actual)
 	}
 	if expected, actual := nodeCreated, node.getState(); expected != actual {

